### PR TITLE
feat: add delete reminder use case #4

### PR DESCRIPTION
**Pull request by oriolagobat**

### CODE REVIEW

File: src/main/kotlin/reminder/application/ReminderDeleter.kt
Comment (Pablito2020) [inline src/main/kotlin/reminder/application/ReminderDeleter.kt:16]: As stated on #4, the use case should:
1) Receive the reminder as a parameter.
2) Search if the reminder exists on the database ((you can use the already done ReminderSearcher use case (see
Use case of searching for reminders #3) )
3) If it doesn't exist throw an exception.
4) Remove the reminder.

The exceptions that the function throws should be stated with the "@throws" keyword in the kdoc (kotlin alternative to javadoc).

File: src/main/kotlin/reminder/domain/ReminderRepository.kt
Comment (Pablito2020) [inline src/main/kotlin/reminder/domain/ReminderRepository.kt:8]: I don't think that returning a boolean is the better choice here, since if it fails (returning false) you don't know which error is happening (is it because an I/O error? Because the reminder doesn't exist? etc).

I think that throwing exceptions for non-desirable cases is a better way to manage the "unusual" cases.

Anyway, I don't know if we should not have any return parameter or returning the deleted reminder should be a better option....

**Review by Pablito2020 — CHANGES_REQUESTED** [review]
Some additional checks that the use case needs and a "discussion" about what should the delete function return.

File: src/main/kotlin/reminder/domain/ReminderRepository.kt
Comment (oriolagobat) [reply src/main/kotlin/reminder/domain/ReminderRepository.kt:8]: I don't think returning the reminder itself is the best way to handle this, as it technically doesn't exist anymore. 
But I agree with the part that making it return `unit` is also not the best practice. 
Given this, what do you think it's better:
- Returning the reminder
- Returning the identifier
- Returning nothing

Comment (Pablito2020) [reply src/main/kotlin/reminder/domain/ReminderRepository.kt:8]: I think that returning the reminder identifier isn't really needed, since you are passing it as a method parameter (so, it means that you already know it).

Returning the reminder itself means that you can do some operations with it (read the title and description about the reminder) that otherwise you couldn't do (since you have just deleted the entry inside the repository, and, as a consequence the getReminder method will return Optional.None).

But the real question here is: Do we really need to check the information of a removed reminder? For now we don't so maybe if the future leads us to this need, we can change this use case and adapt it.

This means that I would opt for returning Unit, and if the needs change, we adapt this use case for our needs.

File: src/main/kotlin/reminder/application/ReminderDeleter.kt
Comment (oriolagobat) [reply src/main/kotlin/reminder/application/ReminderDeleter.kt:16]: Why should it receive the reminder as a parameter, shouldn't it instead receive just the identifier?

Comment (Pablito2020) [reply src/main/kotlin/reminder/application/ReminderDeleter.kt]: Yes, my bad, it should only receive the ReminderIdentifier, but the point I wanted to emphasize was the exception if the reminder doesn't exist.

File: src/test/kotlin/reminder/application/ReminderDeleterTest.kt
Comment (Pablito2020) [inline src/test/kotlin/reminder/application/ReminderDeleterTest.kt:41]: Maybe: "if the reminder exists, it delete its from the repository" would be a better naming? (Yes I am super picky with this stuff 😢 )

**Review by Pablito2020 — CHANGES_REQUESTED** [review]
Wait for the #23 to be merged for doing the next changes (that's why I requested changes), and we'll squash the commits (since this pull request would be 3 commits where only 1 it's needed, the other 2 are because human interaction delay).

**Comment by oriolagobat** [comment]
I'll wait for the approval on #23, then make the changes I've stated on the _TODO_ and we'll merge it then.

**Review by Pablito2020 — APPROVED** [review]

**Review by FerranAD — APPROVED** [review]

**Pull request merged**
